File: metronome/state.py
from dataclasses import dataclass, field


@dataclass
class AppState:
    bpm: int = 120                       # 20–300 inclusive
    beats_per_measure: int = 4           # 1–8 inclusive
    running: bool = False
    tap_timestamps: list[float] = field(default_factory=list)

File: js-metronome/src/state.ts
export interface AppState {
  bpm: number;
  beatsPerMeasure: number;
  running: boolean;
  tapTimestamps: number[];
}

export const BPM_MIN = 20;
export const BPM_MAX = 300;
export const LOOKAHEAD_S = 0.100;
export const SCHEDULE_INTERVAL_MS = 25;
export const FLASH_DURATION_MS = 100;

export function clampBpm(value: number): number {
  return Math.max(BPM_MIN, Math.min(BPM_MAX, Math.round(value)));
}

export function beatInterval(bpm: number): number {
  return 60.0 / bpm;
}

export function isAccent(beatNumber: number, beatsPerMeasure: number): boolean {
  return beatNumber % beatsPerMeasure === 0;
}
